### PR TITLE
Add support for pretraining recurring span selection to Splinter

### PR DESCRIPTION
**Pull request by jvcop**
This pull request aims to add support for recurring span selection pretraining as proposed by the authors of [Splinter](https://arxiv.org/abs/2101.00438). The pretraining objective differs from the question answering task in a couple of ways:

- There is not a question, but a number of question tokens replacing recurring spans.
- The shape of `start_positions` and `end_positions` is `(batch_size, num_questions)` instead of `(batch_size, )`.
- The shape of `start_logits` and `end_logits` is `(batch_size, num_questions, sequence_length)` instead of `(batch_size, sequence_length)`.
- The loss should ignore zero positions, i.e. `ignore_index=0`. Zeros are used in the original code to denote padded question tokens and their start and end positions.

To this end, we added `SplinterForPreTraining`.

Minimal training example:

```python
import torch
from torch.utils.data import IterableDataset
from transformers import SplinterConfig
from transformers import SplinterForPreTraining
from transformers import Trainer
from transformers import TrainingArguments


class QuestionAnsweringDataset(IterableDataset):
    def __iter__(self):
        yield {
            "input_ids": torch.tensor([101, 104, 123, 456, 104, 234, 567, 102]),
            "attention_mask": torch.tensor([1, 1, 1, 1, 1, 1, 1, 1]),
            "token_type_ids": torch.tensor([0, 0, 0, 0, 0, 0, 0, 0]),
            "question_positions": torch.tensor([1, 4]),
            "start_positions": torch.tensor([2, 5]),
            "end_positions": torch.tensor([3, 6]),
        }


config = SplinterConfig()
model = SplinterForPreTraining(config)
dataset = QuestionAnsweringDataset()

trainer = Trainer(
    model=model,
    args=TrainingArguments(max_steps=3, output_dir="/tmp"),
    train_dataset=dataset,
)
trainer.train()
```

CC @tobigue

@patil-suraj @LysandreJik @patrickvonplaten @oriram

### CODE REVIEW

File: src/transformers/models/splinter/modeling_splinter.py [hunk 1]
```diff
@@ -940,3 +941,160 @@ def forward(
...
         )
+
+
+@dataclass
+class SplinterForPreTrainingOutput(ModelOutput):
+    """
+    Class for outputs of Splinter as a span selection model.
+
+    Args:
+        loss (`torch.FloatTensor` of shape `(1,)`, *optional*, returned when `labels` is provided):
```
Comment (patil-suraj) [inline src/transformers/models/splinter/modeling_splinter.py:952]: Nice!

File: src/transformers/models/splinter/modeling_splinter.py [hunk 2]
```diff
@@ -940,3 +941,160 @@ def forward(
...
+            the only one for which start_logits and end_logits are calculated and they will be of shape `(batch_size,
+            sequence_length)`.
+        """
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if question_positions is None:
+            if start_positions is not None and end_positions is not None:
+                raise TypeError(
+                    "For span selection training, question_positions must be specified to obtain the correct loss"
+                )
```
Comment (patil-suraj) [inline src/transformers/models/splinter/modeling_splinter.py:1036]: Since we can prepare `question_positions` using `input_ids`, shouldn't we do this check after that ?

File: src/transformers/models/splinter/modeling_splinter.py [hunk 3]
```diff
@@ -940,3 +941,160 @@ def forward(
...
+            head_mask=head_mask,
+            inputs_embeds=inputs_embeds,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        sequence_output = outputs[0]
+        batch_size, sequence_length, dim = sequence_output.size()
+        start_logits, end_logits = self.splinter_qass(sequence_output, question_positions)  # [B, Q, S]
```
Comment (patil-suraj) [inline src/transformers/models/splinter/modeling_splinter.py:1059]: tiny nit
```suggestion
        start_logits, end_logits = self.splinter_qass(sequence_output, question_positions)  # [batch, num_questions, seq]
```

File: src/transformers/models/splinter/modeling_splinter.py [hunk 4]
```diff
@@ -940,3 +941,160 @@ def forward(
...
+        num_questions = question_positions.size(1)
+        if attention_mask is not None:
+            attention_mask_for_each_question = attention_mask.unsqueeze(1).expand(
+                batch_size, num_questions, sequence_length
+            )
+            start_logits = start_logits + (1 - attention_mask_for_each_question) * -10000.0
+            end_logits = end_logits + (1 - attention_mask_for_each_question) * -10000.0
+
+        total_loss = None
+        if start_positions is not None and end_positions is not None:  # [B, Q, S]
```
Comment (patil-suraj) [inline src/transformers/models/splinter/modeling_splinter.py:1070]: same comment as above.

File: src/transformers/models/splinter/modeling_splinter.py [hunk 5]
```diff
@@ -940,3 +941,160 @@ def forward(
...
+            # sometimes the start/end positions are outside our model inputs, we ignore these terms
+            start_positions.clamp_(0, max(0, sequence_length - 1))
+            end_positions.clamp_(0, max(0, sequence_length - 1))
+
+            # Ignore zero positions in the loss. Splinter never predicts zero
+            # during pretraining and zero is used for padding question
+            # tokens as well as for start and end positions of padded
+            # question tokens.
+            loss_fct = CrossEntropyLoss(ignore_index=0)
```
Comment (patil-suraj) [inline src/transformers/models/splinter/modeling_splinter.py:1079]: Think we could use `config.pad_token_id` instead of hardcoding 0 here.

File: tests/models/splinter/test_modeling_splinter.py [hunk 6]
```diff
@@ -217,3 +332,96 @@ def test_splinter_question_answering(self):
...
+        self.assertEqual(torch.argmax(output.start_logits[0, 1]), 10)
+        self.assertEqual(torch.argmax(output.end_logits[0, 1]), 12)
+
+    @slow
+    def test_splinter_pretraining_loss_requires_question_positions(self):
+        model = SplinterForPreTraining.from_pretrained("tau/splinter-base-qass")
+
+        # Input: "[CLS] [QUESTION] was born in [QUESTION] . Brad returned to the United Kingdom later . [SEP]"
+        # Output should be the spans "Brad" and "the United Kingdom"
+        input_ids = torch.tensor(
```
Comment (patil-suraj) [inline tests/models/splinter/test_modeling_splinter.py:363]: Great tests!

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for adding this. Looks good to me, just left some nits.

File: utils/check_repo.py [hunk 7]
```diff
@@ -185,6 +185,7 @@
     "TFWav2Vec2ForCTC",
     "TFHubertForCTC",
     "MaskFormerForInstanceSegmentation",
+    "SplinterForPreTraining",
```
Comment (sgugger) [inline utils/check_repo.py:188]: Why not add the model to the `AutoModelForPreTraining` class?

**Review by sgugger — APPROVED** [review]
Very nice addition! Just double-checking why we're not adding this model to the pretraining auto class?

File: utils/check_repo.py (hunk 7 above)
Comment (tobigue) [reply utils/check_repo.py]: added

File: src/transformers/models/splinter/modeling_splinter.py (hunk 5 above)
Comment (tobigue) [reply src/transformers/models/splinter/modeling_splinter.py]: done

File: src/transformers/models/splinter/modeling_splinter.py (hunk 4 above)
Comment (tobigue) [reply src/transformers/models/splinter/modeling_splinter.py]: done

File: src/transformers/models/splinter/modeling_splinter.py (hunk 3 above)
Comment (tobigue) [reply src/transformers/models/splinter/modeling_splinter.py]: changed

**Comment by tobigue** [comment]
Thanks for reviewing the PR! :) I added the suggested changes.

@jvcop will add an answer to this https://github.com/huggingface/transformers/pull/17247#discussion_r873691809

File: src/transformers/models/splinter/modeling_splinter.py (hunk 2 above)
Comment (jvcop) [reply src/transformers/models/splinter/modeling_splinter.py:1036]: Good point, I think our original motivation is flawed :) The assumption that there is one question token holds for question answering with Splinter, but not for recurring span selection, as different examples can have a different number of question tokens. This makes the derivation a bit more complicated. I see two options: Either we always raise `TypeError`, effectively requiring `question_positions`, or we add a small snippet for deriving it from `input_ids`. I guess one downside of the first approach is that it would be somewhat weird to require a `typing.Optional`.

Comment (patrickvonplaten) [reply src/transformers/models/splinter/modeling_splinter.py:1036]: IMO the most readable code would be:

```python
if question_positions is None and start_positions is not None and end_positions is not None:
   # loss is calculated, need question positions
   raise 
elif question_positions is None and input_ids is not None:
    # input_ids are required to compute question_positions for forward pass
    raise 
elif question_positions is None and input_ids is not None
    question_positions = ...
```

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me!

File: src/transformers/models/splinter/modeling_splinter.py (hunk 2 above)
Comment (patil-suraj) [reply src/transformers/models/splinter/modeling_splinter.py:1036]: Should be good to merge once this is addressed :)

File: src/transformers/models/splinter/modeling_splinter.py [hunk 8]
```diff
@@ -1029,18 +1029,14 @@ def forward(
...
+        if question_positions is None and start_positions is not None and end_positions is not None:
+            raise TypeError("question_positions must be specified in order to calculate the loss")
 
-            if input_ids is not None:
-                question_positions = torch.argmax(
-                    (torch.eq(input_ids, self.question_token_id)).int(), dim=-1
-                ).unsqueeze(0)
-            else:
-                raise TypeError("question_positions is required when input_embeds is specified")
+        if question_positions is None and input_ids is None:
```
Comment (patrickvonplaten) [inline src/transformers/models/splinter/modeling_splinter.py:1035]:
```suggestion
        elif question_positions is None and input_ids is None:
```

File: src/transformers/models/splinter/modeling_splinter.py [hunk 9]
```diff
@@ -1029,18 +1029,14 @@ def forward(
...
+            raise TypeError("question_positions must be specified when input_embeds is used")
+
+        if question_positions is None:
```
Comment (patrickvonplaten) [inline src/transformers/models/splinter/modeling_splinter.py:1038]:
```suggestion
        elif question_positions is None:
```

Comment (patrickvonplaten) [reply src/transformers/models/splinter/modeling_splinter.py:1038]: Think it's easier to follow the logic like this, but feel free to disregard if you don't agree

**Review by patrickvonplaten — APPROVED** [review]

File: src/transformers/models/splinter/modeling_splinter.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/models/splinter/modeling_splinter.py]: Looks good to me!

**Comment by patrickvonplaten** [comment]
As far as I can see all comments have been addressed - merging! Thanks a lot for your work here @jvcop !

**Comment by jvcop** [comment]
Thanks a lot for the fast review! And to @tobigue who was an integral part of this :tada:
